# Домашнее задание 6. Pandas

**Примечание:**

В заданиях вам даны подсказки по выполнению, a также вывод работы кода авторского решения - вы можете смотреть на этот вывод и понимать, в правильном ли направлении вы движетесь при выполнении задания.

Скачиваем данные:

In [1]:
!curl -L -o 'articles.csv' 'https://drive.google.com/uc?id=1nht5Pd7kuqJSCz7B9jwU4qScnCO1ZPop&export=download&confirm=t'

curl: (3) URL rejected: Port number was not a decimal number between 0 and 65535
"export" ­Ґ пў«пҐвбп ў­гв७­Ґ© Ё«Ё ў­Ґи­Ґ©
Є®¬ ­¤®©, ЁбЇ®«­пҐ¬®© Їа®Ја ¬¬®© Ё«Ё Ї ЄҐв­л¬ д ©«®¬.
"confirm" ­Ґ пў«пҐвбп ў­гв७­Ґ© Ё«Ё ў­Ґи­Ґ©
Є®¬ ­¤®©, ЁбЇ®«­пҐ¬®© Їа®Ја ¬¬®© Ё«Ё Ї ЄҐв­л¬ д ©«®¬.


In [2]:
import pandas as pd
import numpy as np

# Задача 0

Загрузите таблицу с данными из `articles.csv`:


In [3]:
articles = pd.read_csv('C:\\Users\\A\\articles.csv')

Посмотрите на первые 5 строк датафрейма (метод `head`):

In [4]:
articles.head(10)

id                                              title     publication  \
0  17283  House Republicans Fret About Winning Their Hea...  New York Times   
1  17284  Rift Between Officers and Residents as Killing...  New York Times   
2  17285  Tyrus Wong, ‘Bambi’ Artist Thwarted by Racial ...  New York Times   
3  17286  Among Deaths in 2016, a Heavy Toll in Pop Musi...  New York Times   
4  17287  Kim Jong-un Says North Korea Is Preparing to T...  New York Times   
5  17288  Sick With a Cold, Queen Elizabeth Misses New Y...  New York Times   
6  17289  Taiwan’s President Accuses China of Renewed In...  New York Times   
7  17290  After ‘The Biggest Loser,’ Their Bodies Fought...  New York Times   
8  17291  First, a Mixtape. Then a Romance. - The New Yo...  New York Times   
9  17292  Calling on Angels While Enduring the Trials of...  New York Times   

                          author        date  year  month  \
0                     Carl Hulse  2016-12-31  2016     12   
1  Benjamin Mueller and Al Baker  2017-06-19  2017      6   
2                   Margalit Fox  2017-01-06  2017      1   
3               William McDonald  2017-04-10  2017      4   
4                  Choe Sang-Hun  2017-01-02  2017      1   
5                    Sewell Chan  2017-01-02  2017      1   
6            Javier C. Hernández  2017-01-02  2017      1   
7                    Gina Kolata  2017-02-08  2017      2   
8               Katherine Rosman  2016-12-31  2016     12   
9                    Andy Newman  2016-12-31  2016     12   

                                             content  
0  WASHINGTON  —   Congressional Republicans have...  
1  After the bullet shells get counted, the blood...  
2  When Walt Disney’s “Bambi” opened in 1942, cri...  
3  Death may be the great equalizer, but it isn’t...  
4  SEOUL, South Korea  —   North Korea’s leader, ...  
5  LONDON  —   Queen Elizabeth II, who has been b...  
6  BEIJING  —   President Tsai   of Taiwan sharpl...  
7  Danny Cahill stood, slightly dazed, in a blizz...  
8  Just how   is Hillary Kerr, the    founder of ...  
9  Angels are everywhere in the Muñiz family’s ap...

```
	id	title	publication	author	date	year	month	content
0	17283	House Republicans Fret About Winning Their Hea...	New York Times	Carl Hulse	2016-12-31	2016	12	WASHINGTON — Congressional Republicans have...
1	17284	Rift Between Officers and Residents as Killing...	New York Times	Benjamin Mueller and Al Baker	2017-06-19	2017	6	After the bullet shells get counted, the blood...
2	17285	Tyrus Wong, ‘Bambi’ Artist Thwarted by Racial ...	New York Times	Margalit Fox	2017-01-06	2017	1	When Walt Disney’s “Bambi” opened in 1942, cri...
3	17286	Among Deaths in 2016, a Heavy Toll in Pop Musi...	New York Times	William McDonald	2017-04-10	2017	4	Death may be the great equalizer, but it isn’t...
4	17287	Kim Jong-un Says North Korea Is Preparing to T...	New York Times	Choe Sang-Hun	2017-01-02	2017	1	SEOUL, South Korea — North Korea’s leader, ...
```

Определите количество строк и столбцов в датафрейме (атрибут `shape`)

In [21]:
articles.shape

(50000, 8)

`(50000, 8)`

Посмотрите, в каких столбцах есть пропуски (используйте методы ```isnull``` и ```sum```)

In [37]:
pd.isnull(articles).sum()

id                0
title             0
publication       0
author         6306
date              0
year              0
month             0
content           0
dtype: int64

```
id                0
title             0
publication       0
author         6306
date              0
year              0
month             0
content           0
dtype: int64
```

Видим, что в колонке ```author``` есть пропуски.

Посчитайте, какая доля пропусков содержится в этой колонке (используйте методы ```isnull``` и ```mean```)

In [38]:
pd.isnull(articles).mean().author

0.12612

`0.12612`

Видим, что в колонке ```author``` 12% пропусков. Давайте удалим те строки, в которых пропущено значение в колонке ```author```

Для этого используйте функцию ```dropna```, которой в качестве аргумента можно передать параметр ```axis```, чтобы удалить строки или столбцы с пропусками.

Не забудьте, что чтобы сохранить результат удаления строк, нужно либо перезаписать датафрейм (```df = df.same_method()```), либо выставить в методе ```dropna``` параметр ```inplace=True```, который говорит о том, что операция будет производиться с самим датафреймом, а не с его копией


In [43]:
articles.dropna(inplace=True)

Проверим размер датафрейма после удаления пропусков и убедимся, что в датафрейме теперь нет пропусков в данных

In [48]:
articles.shape

(43694, 8)

`(43694, 8)`

In [49]:
pd.isnull(articles).sum()

id             0
title          0
publication    0
author         0
date           0
year           0
month          0
content        0
dtype: int64

```
id             0
title          0
publication    0
author         0
date           0
year           0
month          0
content        0
dtype: int64
```

Проделайте следующие базовые операции с датафреймами:

# Задача 1


Определите количество различных издательств в таблице и их имена (используйте для этого методы датафрейма ```unique``` или ```nunique```):

In [54]:
articles.publication.nunique()

5

`5`

In [55]:
articles.publication.unique()

array(['New York Times', 'Breitbart', 'CNN', 'Business Insider',
       'Atlantic'], dtype=object)

```
array(['New York Times', 'Breitbart', 'CNN', 'Business Insider',
       'Atlantic'], dtype=object)
```

# Задача 2

Найдите количество опубликованных статей в отрезке с ```2016-06-03``` по ```2016-12-07``` (включая эти дни)

Для этого используйте фильтрацию по условию. С датами можно обращаться просто как со строками.

При желании это все можно написать одной строкой.

**Подсказка:** найдите строки c датой публикации ```>=``` стартовой даты и получите булеву маску. Аналогично получите булеву маску для второй даты. Затем с помощью операции ```&``` над двумя булевыми масками получите новый булев массив, в котором ```True``` будет стоять только у тех строк, у которых дата публикации попадает в желаемый отрезок. А затем используйте эту булеву маску как индексы датафрейма.

In [60]:
articles[(articles.date >= '2016-06-03') & (articles.date <= '2016-12-07')].shape[0]

15079

`15079`

# Задача 3

Какое издание выпустило больше всего статей в 2016 году?

In [73]:
articles[articles['year'] == 2016]['publication'].value_counts().idxmax()

'Breitbart'

`'Breitbart'`

# Задача 4

Посчитайте количество статей автора *Tom Ciccotta* в каждом году.

Для этого сначала отберите строки, в которых автором является *Tom Ciccotta*, а затем создайте счетчик из столбца ```year``` (для этого просто позовите метод ```value_counts()``` от столбца ```year```)

In [5]:
articles[articles['author'] == 'Tom Ciccotta'].value_counts('year')

year
2017    132
2016    124
Name: count, dtype: int64

```
2017    132
2016    124
Name: year, dtype: int64
```

## Задача 5

Выпишите названия 3х первых статей автора *John Hayward* в 2016 году.

**Подсказка 1:** после отбора статей по автору и году, не забудте отсортировать датафрейм по столбцу даты публикации.

**Подсказка 2:** срезы по строкам датафреймов работают так же, как срезы по спискам.

In [108]:
articles[(articles['author'] == 'John Hayward') & (articles['year'] == 2016)].sort_values(by='date')['title'].head(3).values

array(['Indonesian Couple Beaten with Canes for Violating Sharia Law',
       'Islamic State Claims Credit For Gun Attack On Russian Tourists',
       'Protests Across India Against Saudi Execution of Shiite Cleric'],
      dtype=object)

```
array(['Indonesian Couple Beaten with Canes for Violating Sharia Law',
       'Islamic State Claims Credit For Gun Attack On Russian Tourists',
       'Protests Across India Against Saudi Execution of Shiite Cleric'],
      dtype=object)
```

# Задача 6

Какой автор на сотом месте по количеству опубликованных статей? Сколько статей он опубликовал?

**Подсказка:** для подсчета встречаемости значений используйте метод ```value_counts```. Для доступа к индексам полученного после ```value_counts``` ```pd.Series``` используйте атрибут ```index```

In [4]:
articles['author'].value_counts().index[99]

'Milo'

`'Milo'`

In [3]:
articles[articles['author'] == 'Milo'].count()['title']

86

`86`

# Работа со строками в датафрейме

Еще одной важной частью работы с данными является **нормализация данных**.

Это особенно актуально при работе с данными в текством формате и при работе с данными из нескольких разнородных источников. Например, один и тот же номер телефона пользователя может записан в базах данных в разных форматах:

- `+7(918)123-45-67`
- `+7(918)1234567`
- `+79181234567`
- `+8(918)123-45-67`
- `8(918)1234567`
- `89181234567`

Поэтому важно нормализовывать данные, чтобы иметь возможность объединить все данные, которые относятся к одному реальному объекту.

Именно нормализации данных при работе со строками будем учиться в этой части домашней работы.

Для датафреймов существуют методы работы со строковыми данными. Чтобы применить их, необходимо воспользоваться атрибутом ```str```, после чего вызвать нужные методы работы со строками. Например, вызов:

```df['content'].str.len() ```

подсчитает для каждой строчки в датафрейме количество символов в колонке content. Примеры кода по работе с текстовыми данными в pandas можно найти в гайде [working with text data](https://pandas.pydata.org/pandas-docs/stable/user_guide/text.html) документации, там же вы найдете список методов пандаса по работе с текстовыми данными.


# Задача 7

Найдите в датафрейме всех авторов, имя которых содержит ```Faith```. Выведите Series, состоящий из всех таких уникальных имен.

In [62]:
pd.Series(articles[articles['author'].str.contains('Faith', na=False)]['author'].unique())

0     Faith Haleh Robinson
1             Faith Karimi
2     Faith Haleh Robinson
3            Faith Karimi,
4            Faith Karimi 
5           Faith Karimi, 
6             Faith Karimi
7            Faith Karimi 
dtype: object

```
0     Faith Haleh Robinson
1             Faith Karimi
2     Faith Haleh Robinson
3            Faith Karimi,
4            Faith Karimi
5           Faith Karimi,
6             Faith Karimi
7            Faith Karimi
dtype: object
```

Как можно заметить, в таблице существует множество различных написаний имени Faith Karimi. В основном эти написания различаются пунктуацией - лишние пробелы и запятые. Для правильного подсчета статистик для текстовых данных зачастую возникает необходимость в их предобработке.

**Методы строк** можно вспомнить по этой ссылке: https://pythonworld.ru/tipy-dannyx-v-python/stroki-funkcii-i-metody-strok.html

Проведите следующие преобразования для колонок ```author``` и ```content```:

# Задача 8

Приведение текста к нижнему регистру:

In [71]:
articles['author'] = articles['author'].str.lower()
articles['content'] = articles['content'].str.lower()

In [72]:
articles

id                                              title  \
0      17283  House Republicans Fret About Winning Their Hea...   
1      17284  Rift Between Officers and Residents as Killing...   
2      17285  Tyrus Wong, ‘Bambi’ Artist Thwarted by Racial ...   
3      17286  Among Deaths in 2016, a Heavy Toll in Pop Musi...   
4      17287  Kim Jong-un Says North Korea Is Preparing to T...   
...      ...                                                ...   
49995  73465   Rex Tillerson Says Climate Change Is Real, but …   
49996  73466  The Biggest Intelligence Questions Raised by t...   
49997  73467  Trump Announces Plan That Does Little to Resol...   
49998  73468    Dozens of For-Profit Colleges Could Soon Close    
49999  73469                       The Milky Way’s Stolen Stars   

          publication                         author        date  year  month  \
0      New York Times                     carl hulse  2016-12-31  2016     12   
1      New York Times  benjamin mueller and al baker  2017-06-19  2017      6   
2      New York Times                   margalit fox  2017-01-06  2017      1   
3      New York Times               william mcdonald  2017-04-10  2017      4   
4      New York Times                  choe sang-hun  2017-01-02  2017      1   
...               ...                            ...         ...   ...    ...   
49995        Atlantic                 robinson meyer  2017-01-11  2017      1   
49996        Atlantic                     amy zegart  2017-01-11  2017      1   
49997        Atlantic                  jeremy venook  2017-01-11  2017      1   
49998        Atlantic                    emily deruy  2017-01-11  2017      1   
49999        Atlantic                   marina koren  2017-01-11  2017      1   

                                                 content  
0      washington  —   congressional republicans have...  
1      after the bullet shells get counted, the blood...  
2      when walt disney’s “bambi” opened in 1942, cri...  
3      death may be the great equalizer, but it isn’t...  
4      seoul, south korea  —   north korea’s leader, ...  
...                                                  ...  
49995  as chairman and ceo of exxonmobil, rex tillers...  
49996  i’ve spent nearly 20 years looking at intellig...  
49997    donald trump will not be taking necessary st...  
49998  dozens of   colleges could be forced to close ...  
49999  the force of gravity can be described using a ...  

[50000 rows x 8 columns]

```
     id	title	publication	author	date	year	month	content
0	17283	House Republicans Fret About Winning Their Hea...	New York Times	carl hulse	2016-12-31	2016	12	washington — congressional republicans have...
1	17284	Rift Between Officers and Residents as Killing...	New York Times	benjamin mueller and al baker	2017-06-19	2017	6	after the bullet shells get counted, the blood...
2	17285	Tyrus Wong, ‘Bambi’ Artist Thwarted by Racial ...	New York Times	margalit fox	2017-01-06	2017	1	when walt disney’s “bambi” opened in 1942, cri...
3	17286	Among Deaths in 2016, a Heavy Toll in Pop Musi...	New York Times	william mcdonald	2017-04-10	2017	4	death may be the great equalizer, but it isn’t...
4	17287	Kim Jong-un Says North Korea Is Preparing to T...	New York Times	choe sang-hun	2017-01-02	2017	1	seoul, south korea — north korea’s leader, ...
```

# Задача 9

Удаление всей пунктуации из текста.

**Подсказка:** для вас уже создано начальное множество символов (достаточно ли его?). Символы надо исключить с помощью метода ```replace```.

Данного множества недостаточно, для примера, не хватает следующих символов: ’, ” и тд.

In [77]:
import string
punctuation = set(string.punctuation)
additional_punctuation = {"’", "”", "“", "-", "—"}
punctuation.update(additional_punctuation)

In [78]:
# может работать несколько секунд
def remove_punct(text):
    for char in punctuation:
        text = text.replace(char, "")
    return text

articles['content'] = articles['content'].apply(remove_punct)

In [79]:
# пример того, что должно получиться после этого шага
articles.content.loc[45926]

'back in the late 90s getting a startup big enough to get onto the stock market was the thing to do  now a market crash a great recession and a possible   burst later     investors are  and the public stock market is increasingly volatile still going public is often viewed as the end goal for ambitious young tech companies  heres how and why startups wind up on the stock market     and why  back in the late 90s getting a startup big '

```
'back in the late 90s getting a startup big enough to get onto the stock market was the thing to do  now a market crash a great recession and a possible burst later  investors are and the public stock market is increasingly volatile still going public is often viewed as the end goal for ambitious young tech companies  heres how and why startups wind up on the stock market  and why  back in the late 90s getting a startup big '
```

# Задача 10

Удаление пробелов в начале и конце строки (воспользуйтесь готовым методом):

In [80]:
articles['content'] = articles['content'].str.strip()

In [81]:
# check
articles.content.loc[45926]

'back in the late 90s getting a startup big enough to get onto the stock market was the thing to do  now a market crash a great recession and a possible   burst later     investors are  and the public stock market is increasingly volatile still going public is often viewed as the end goal for ambitious young tech companies  heres how and why startups wind up on the stock market     and why  back in the late 90s getting a startup big'

```
'back in the late 90s getting a startup big enough to get onto the stock market was the thing to do  now a market crash a great recession and a possible burst later  investors are and the public stock market is increasingly volatile still going public is often viewed as the end goal for ambitious young tech companies  heres how and why startups wind up on the stock market  and why  back in the late 90s getting a startup big'
```

# Задача 11

Замена подряд идущих пробелов одним пробелом:

In [83]:
articles['content'] = articles['content'].str.replace(r'\s+', ' ', regex=True)

In [84]:
# check
articles.content.loc[45926]

'back in the late 90s getting a startup big enough to get onto the stock market was the thing to do now a market crash a great recession and a possible burst later investors are and the public stock market is increasingly volatile still going public is often viewed as the end goal for ambitious young tech companies heres how and why startups wind up on the stock market and why back in the late 90s getting a startup big'

```
'back in the late 90s getting a startup big enough to get onto the stock market was the thing to do now a market crash a great recession and a possible burst later investors are and the public stock market is increasingly volatile still going public is often viewed as the end goal for ambitious young tech companies heres how and why startups wind up on the stock market and why back in the late 90s getting a startup big'
```

Если последовательно применить код из задач 8-11 к строке ```"   It's 6 a.m. and I'm still doing this homework :((    "```, то она преобразуется в строку  ```its 6 am and im still doing this homework```

In [89]:
text = pd.Series(["   It's 6 a.m. and I'm still doing this homework :((    "])

text = text.str.lower()

punctuation = set(string.punctuation)
additional_punctuation = {"’", "”", "“", "-", "—"}
punctuation.update(additional_punctuation)

def remove_punct(text):
    for char in punctuation:
        text = text.replace(char, "")
    return text

text = text.apply(remove_punct)

text = text.str.strip()

text = text.str.replace(r'\s+', ' ', regex=True)

text.iloc[0]

'its 6 am and im still doing this homework'


Снова найдите в датафрейме всех авторов, имя которых содержит faith. Проверьте, что теперь различные способы написания «схлопываются» в один.

In [102]:
authors_with_faith = articles[articles['author'].str.contains('faith', na=False)]['author']

authors_cleaned = authors_with_faith.str.lower().apply(remove_punct).str.strip().str.replace(r'\s+', ' ', regex=True)

unique_authors = authors_cleaned.unique()

In [104]:
pd.Series(unique_authors)

0    faith haleh robinson
1            faith karimi
dtype: object

```
0    faith haleh robinson
1            faith karimi
dtype: object
```

Еще задачи:

- https://www.machinelearningplus.com/python/101-pandas-exercises-python/
- https://www.kaggle.com/code/tangchengshun/pandas-75-exercises-with-solutions
- https://pandaspractice.com/
- https://www.w3resource.com/python-exercises/pandas/index.php
- https://grossmend.com/blog/post/pandas_120_part_2/